src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,169 @@
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of bound ");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+
+        Node<T> newNode = new Node<>(value);
+        if (index == 0) {
+            newNode.next = first;
+            first.previous = newNode;
+            first = newNode;
+        } else {
+            Node<T> indexNode = first;
+            for (int i = 0; i < index; i++) {
+                indexNode = indexNode.next;
+            }
+            newNode.previous = indexNode.previous;
+            newNode.next = indexNode;
+            indexNode.previous.next = newNode;
+            indexNode.previous = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T sample : list) {
+            add(sample);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of bound ");
+        }
+        Node<T> currentNode = first;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode.element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of bound ");
+        }
+        Node<T> currentNode = first;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        T oldValue = currentNode.element;
+        currentNode.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of bound ");
+        }
+        Node<T> nodeToRemove;
+        if (index == 0) {
+            nodeToRemove = first;
+            first = first.next;
+            if (first != null) {
+                first.previous = null;
+            }
+            if (first == null) {
+                last = null;
+            }
+        } else if (index == size - 1) {
+            nodeToRemove = last;
+            last = last.previous;
+            if (last != null) {
+                last.next = null;
+            }
+            if (last == null) {
+                first = null;
+            }
+        } else {
+            nodeToRemove = first;
+            for (int i = 0; i < index; i++) {
+                nodeToRemove = nodeToRemove.next;
+            }
+            nodeToRemove.previous.next = nodeToRemove.next;
+            nodeToRemove.next.previous = nodeToRemove.previous;
+        }
+        size--;
+        return nodeToRemove.element;
```

Comment:
Create a separate method `unlink(Node node)` for removing a node from the list. This logic will be used in two methods: when we remove from the list by index and by value. So let's create a method that will take a Node that needs to be removed and unlink it, and call it from both methods when Node is found.